theseus/optimizer/nonlinear/nonlinear_optimizer.py

```diff
@@ -191,43 +196,132 @@ def _optimize_impl(
+                info=info,
+                track_best_solution=track_best_solution,
+                verbose=verbose,
+                force_update=False,
+                **kwargs,
+            )
+        elif backward_mode in [BackwardMode.IMPLICIT, BackwardMode.TRUNCATED]:
+            if backward_mode == BackwardMode.IMPLICIT:
+                backward_num_iterations = 1
+            else:
+                if "backward_num_iterations" not in kwargs:
+                    raise ValueError(
+                        "backward_num_iterations expected but not received"
+                    )
+                backward_num_iterations = kwargs["backward_num_iterations"]
+
+            num_no_grad_iter = self.params.max_iterations - backward_num_iterations
+            with torch.no_grad():
+                self._optimize_loop(
+                    start_iter=0,
+                    num_iter=num_no_grad_iter,
+                    info=info,
+                    track_best_solution=track_best_solution,
+                    verbose=verbose,
+                    force_update=False,
+                    **kwargs,
+                )
+
+            grad_loop_info = self._init_info(track_best_solution, verbose)
+            self._optimize_loop(
+                start_iter=0,
+                num_iter=backward_num_iterations,
+                info=grad_loop_info,
+                track_best_solution=False,
+                verbose=verbose,
+                force_update=True,
+                **kwargs,
+            )
+
+            # Merge the converged status into the info from the detached loop,
+            # and for now, don't update the best err tracking or best solution.
+            M = info.status == NonlinearOptimizerStatus.MAX_ITERATIONS
+            assert np.all(
+                (grad_loop_info.status[M] == NonlinearOptimizerStatus.MAX_ITERATIONS)
+                | (grad_loop_info.status[M] == NonlinearOptimizerStatus.CONVERGED)
+            )
+            info.status[M] = grad_loop_info.status[M]
+            info.converged_iter[M] = (
+                info.converged_iter[M] + grad_loop_info.converged_iter[M]
+            )
+
+            return info
+        else:
+            raise ValueError("Unrecognized backward mode")
+
     @abc.abstractmethod
     def compute_delta(self, **kwargs) -> torch.Tensor:
         pass
 
     # retracts all variables in the given order and updates their values
     # with the result
     def retract_and_update_variables(
-        self, delta: torch.Tensor, converged_indices: torch.Tensor
+        self,
+        delta: torch.Tensor,
+        converged_indices: torch.Tensor,
+        force_update: bool = False,
     ):
         var_idx = 0
         delta = self.params.step_size * delta
         for var in self.linear_solver.linearization.ordering:
             new_var = var.retract(delta[:, var_idx : var_idx + var.dof()])
             var.update(new_var.data, batch_ignore_mask=converged_indices)
```

Comment:
Should this line be removed? (since it's repeated inside the `else` clause)

Reply:
Ah good catch, I removed it